client/client.go

```diff
@@ -22,57 +22,334 @@ limitations under the License.
-	Id      string
-	Status  string
-	Created int64
-	Image   string
+// DockerClientInterface provides methods i.a. for interaction with the docker API.
+type DockerClientInterface interface {
+	ListContainersAsMap() (map[string]docker.APIContainers, error)
+	GetStatsFromContainer(string, bool) (*wrapper.Statistics, error)
+	InspectContainer(string) (*docker.Container, error)
+	FindCgroupMountpoint(string) (string, error)
 }
 
-type DockerClientInterface interface {
-	ListContainers() ([]ContainerInfo, error)
-	FindCgroupMountpoint(subsystem string) (string, error)
+// DockerClient
+type DockerClient struct {
+	cl                 *docker.Client
+	inspectCache       map[string]*docker.Container
+	inspectMutex       sync.Mutex
+	diskUsageCollector fs.DiskUsageCollector
 }
 
-type dockerClient struct{}
+type deviceInfo struct {
+	device string
+	major  string
+	minor  string
+}
 
-func NewDockerClient() *dockerClient {
-	return new(dockerClient)
+// NewDockerClient returns dockerClient instance ready for communication with the server endpoint `unix:///var/run/docker.sock`
+func NewDockerClient() (*DockerClient, error) {
+	client, err := docker.NewClient(endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot initialize docker client instance with the given server endpoint `%s`, err=%v", endpoint, err)
+	}
+
+	dc := &DockerClient{
+		cl:                 client,
+		inspectCache:       map[string]*docker.Container{},
+		diskUsageCollector: fs.DiskUsageCollector{},
+	}
+
+	dc.diskUsageCollector.Init()
+	// get version of docker engine
+	version, err := dc.version()
+	if err != nil {
+		return nil, err
+	}
+	config.DockerVersion = version
+
+	return dc, nil
 }
 
-func (dc *dockerClient) FindCgroupMountpoint(subsystem string) (string, error) {
+// FindCgroupMountpoint returns cgroup mountpoint of a given subsystem
+func (dc *DockerClient) FindCgroupMountpoint(subsystem string) (string, error) {
 	return cgroups.FindCgroupMountpoint(subsystem)
 }
 
-func (dc *dockerClient) ListContainers() ([]ContainerInfo, error) {
-	cl, err := docker.NewClient(endpoint)
+// GetShortID returns short container ID (12 chars)
+func GetShortID(dockerID string) (string, error) {
+	if len(dockerID) < 12 {
+		return "", fmt.Errorf("Docker id %v is too short (the length of id should equal at least 12)", dockerID)
+	}
+	return dockerID[:12], nil
+}
+
+// GetStatsFromContainer returns docker containers stats: cgroups stats (cpu usage, memory usage, etc.) and network stats (tx_bytes, rx_bytes etc.);
+// notes that incoming container id has to be full-length to be able to inspect container
+func (dc *DockerClient) GetStatsFromContainer(id string, collectFs bool) (*wrapper.Statistics, error) {
+	var (
+		err        error
+		pid        int
+		workingSet uint64
+
+		container = &docker.Container{}
+		groupWrap = wrapper.Cgroups2Stats // wrapper for cgroup name and interface for stats extraction
+		stats     = wrapper.NewStatistics()
+	)
+
+	if !isHost(id) {
```

Comment:
I agree that adding more logs would be helpful in debugging. Hovewer, I would like to avoid overloaded with logs in that form (fmt.Fprintf). Probably logrus with an appropriate fields could be the best solution to help tracing collection process. All docker plugin files should be adjusted to this change to keep consistency. I created an issue to address that enhancement (https://github.com/intelsdi-x/snap-plugin-collector-docker/issues/34)